pkg/migration/dryrun.go

```diff
@@ -0,0 +1,263 @@
+		vol := v.(v1.Volume)
+
+		if vol.HostPath != nil {
+			return vol.HostPath.Path
+		}
+
+		return vol.Name
+	}
+
+	getVolumeList := func(arr []v1.Volume) []interface{} {
+		var objs []interface{}
+		for _, obj := range arr {
+			// "crioconf" only exists with operator install.
+			// "dev" only exists with daemonset install.
+			if obj.Name == "crioconf" || obj.Name == "dev" {
+				continue
+			}
+			objs = append(objs, obj)
+		}
+		return objs
+	}
+
+	return util.DeepEqualObjects(
+		getVolumeList(vol1),
+		getVolumeList(vol2),
+		getVolumeKey,
+		h.deepEqualVolume)
+}
+
+func (h *Handler) deepEqualVolume(obj1, obj2 interface{}) error {
+	t1 := obj1.(v1.Volume)
+	t2 := obj2.(v1.Volume)
+
+	vol1 := t1.DeepCopy()
+	vol2 := t2.DeepCopy()
+
+	cleanupFields := func(vol v1.Volume) v1.Volume {
+		// We dont compare name, but only the content
+		// Ignore hostPath type as nil may represent default value.
+		vol.Name = ""
+		if vol.HostPath != nil {
+			vol.HostPath.Type = nil
+		}
+		return vol
+	}
+
+	return util.DeepEqualObject(cleanupFields(*vol1), cleanupFields(*vol2))
+}
+
+func (h *Handler) deepEqualContainers(c1, c2 []v1.Container) error {
+	getContainerName := func(v interface{}) string {
+		return v.(v1.Container).Name
+	}
+	getContainerList := func(arr []v1.Container) []interface{} {
+		objs := make([]interface{}, len(arr))
+		for i, obj := range arr {
+			objs[i] = obj
+		}
+		return objs
+	}
+
+	return util.DeepEqualObjects(
+		getContainerList(c1),
+		getContainerList(c2),
+		getContainerName,
+		h.deepEqualContainer)
+}
+
+// DeepEqualContainer compare two containers
+func (h *Handler) deepEqualContainer(obj1, obj2 interface{}) error {
+	t1 := obj1.(v1.Container)
+	t2 := obj2.(v1.Container)
+
+	c1 := t1.DeepCopy()
+	c2 := t2.DeepCopy()
+
+	msg := ""
+	if c1.Image != c2.Image {
+		msg += fmt.Sprintf("image is different: %s, %s\n", c1.Image, c2.Image)
+	}
+
+	sort.Strings(c1.Command)
+	sort.Strings(c2.Command)
+	if !reflect.DeepEqual(c1.Command, c2.Command) {
+		msg += fmt.Sprintf("command is different: %s, %s\n", t1.Command, t2.Command)
+	}
+
+	sort.Strings(c1.Args)
+	sort.Strings(c2.Args)
+	if !reflect.DeepEqual(c1.Args, c2.Args) {
+		msg += fmt.Sprintf("args is different: %s, %s\n", t1.Args, t2.Args)
+	}
+
+	err := h.deepEqualEnvVars(c1.Env, c2.Env)
+	if err != nil {
+		msg += fmt.Sprintf("env vars is different, %v\n", err)
+	}
+
+	err = h.deepEqualVolumeMounts(c1.VolumeMounts, c2.VolumeMounts)
+	if err != nil {
+		msg += fmt.Sprintf("VolumeMounts is different, %v\n", err)
+	}
+
+	if msg != "" {
+		return fmt.Errorf(msg)
+	}
+	return nil
+}
+
+func (h *Handler) deepEqualEnvVars(env1, env2 []v1.EnvVar) error {
+	getName := func(v interface{}) string {
+		return v.(v1.EnvVar).Name
+	}
+	getList := func(arr []v1.EnvVar) []interface{} {
+		var objs []interface{}
+		skippedEnvs := []string{
+			"PX_TEMPLATE_VERSION",
+			"PORTWORX_CSIVERSION",
+			"CSI_ENDPOINT",
+			"NODE_NAME",
+			"PX_NAMESPACE",
+			"PX_SECRETS_NAMESPACE",
+		}
```

Comment:
We could just create a simple `map[string]struct{}`. Then we don't need to iterate over this slice again and again.

Reply:
resolved